lib/rules/no-restricted-imports.js

```diff
@@ -114,38 +132,132 @@ module.exports = {
+        function isRestrictedForEverythingImported(importSource, importNames) {
+            return Object.prototype.hasOwnProperty.call(restrictedPathMessages, importSource) &&
+                restrictedPathMessages[importSource].importNames &&
+                isEverythingImported(importNames);
+        }
+
+        /**
+         * Check if the given importNames are restricted given a list of restrictedImportNames.
+         * @param {Set.<string>} importNames - Set of import names that are being imported
+         * @param {[string]} restrictedImportNames - array of import names that are restricted for this import
+         * @returns {boolean} whether the objectName is restricted
+         * @private
+         */
+        function isRestrictedObject(importNames, restrictedImportNames) {
+            return restrictedImportNames.some(restrictedObjectName => (
+                importNames.has(restrictedObjectName)
+            ));
+        }
+
+        /**
+         * Check if the given importSource is a restricted path.
+         * @param {string} importSource - path of the import
+         * @param {Set.<string>} importNames - Set of import names that are being imported
          * @returns {boolean} whether the variable is a restricted path or not
          * @private
          */
-        function isRestrictedPath(name) {
-            return Object.prototype.hasOwnProperty.call(restrictedPathMessages, name);
+        function isRestrictedPath(importSource, importNames) {
+            let isRestricted = false;
+
+            if (Object.prototype.hasOwnProperty.call(restrictedPathMessages, importSource)) {
+                if (restrictedPathMessages[importSource].importNames) {
+                    isRestricted = isRestrictedObject(importNames, restrictedPathMessages[importSource].importNames);
+                } else {
+                    isRestricted = true;
+                }
+            }
+
+            return isRestricted;
+        }
+
+        /**
+         * Check if the given importSource is restricted by a pattern.
+         * @param {string} importSource - path of the import
+         * @returns {boolean} whether the variable is a restricted pattern or not
+         * @private
+         */
+        function isRestrictedPattern(importSource) {
+            return restrictedPatterns.length > 0 && restrictedPatternsMatcher.ignores(importSource);
         }
 
         return {
             ImportDeclaration(node) {
+
                 if (node && node.source && node.source.value) {
-                    const importName = node.source.value.trim();
+                    const importSource = node.source.value.trim();
+                    const importNames = node.specifiers.reduce((set, specifier) => {
+                        if (specifier && specifier.type === "ImportDefaultSpecifier") {
+                            set.add("default");
+                        } else if (specifier && specifier.type === "ImportNamespaceSpecifier") {
```

Comment:
Are the two null checks on lines 242 and 244 necessary? It seems like `specifier` will always be non-null.

Reply:
Agreed.  These shouldn't be necessary, removing them with the next commit.